src/ApparentHorizons/HorizonManagerComponentActions.hpp

```diff
@@ -0,0 +1,597 @@
+        "{%s,%s}: send_data_to_ah_finder: have %d points to send\n\n",
+        Parallel::my_node(), Parallel::my_proc(), AhTag::label(), timestep,
+        n_pts);
+  }
+
+  if (not interp_info.global_offsets.empty()) {
+    auto& receiver_proxy =
+        Parallel::get_parallel_component<ah::Finder<Metavariables, AhTag>>(
+            cache);
+    Parallel::simple_action<Actions::Finder::ReceiveInterpolatedVars<AhTag>>(
+        receiver_proxy, interp_info.vars, interp_info.global_offsets);
+  }
+
+  // Flag that the data has been sent, and clear the arrays that we
+  // don't need anymore so we free up memory.
+  db::mutate<Tags::HorizonHolders<Metavariables>>(
+      make_not_null(&box),
+      [&timestep](const gsl::not_null<
+                  db::item_type<Tags::HorizonHolders<Metavariables>>*>
+                      holders_l) {
+        auto& interp_info_l = get<HorizonHolderTag<AhTag>>(*holders_l)
+                                  .interpolated_vars_info.at(timestep);
+        interp_info_l.data_has_been_sent = true;
+        interp_info_l.vars.clear();
+        interp_info_l.global_offsets.clear();
+      });
+}
+
+// Try to interpolate the data it has, and if successful,
+// send interpolated data to the horizon finder.
+template <typename Metavariables, typename DbTags, typename AhTag>
+void try_to_interpolate_data(db::DataBox<DbTags>& box,
+                             Parallel::ConstGlobalCache<Metavariables>& cache,
+                             const Time& timestep) {
+  const auto& holders = db::get<Tags::HorizonHolders<Metavariables>>(box);
+  const auto& interp_vars_info =
+      get<HorizonHolderTag<AhTag>>(holders).interpolated_vars_info;
+
+  // If we don't yet have any points for this HorizonInfo,
+  // we should exit (we can't interpolate anyway).
+  // Also, if we have already reduced data for this HorizonInfo, we should
+  // exit (we don't want to send the HorizonChare the same
+  // information twice).
+  if (interp_vars_info.find(timestep) == interp_vars_info.end() or
+      interp_vars_info.at(timestep).data_has_been_sent) {
+    return;
+  }
+
+  interpolate_data<Metavariables, DbTags, AhTag>(box, timestep);
+
+  // Reduce the horizon data only if all of the elements have interpolated
+  // onto it.
+  const auto& verbose = db::get<::Tags::Verbosity>(box);
+  const auto& num_elements = db::get<Tags::NumElements>(box);
+  if (verbose == ::Verbosity::Debug) {
+    Parallel::printf(
+        "### Node:%d  Proc:%d ###\n"
+        "{%s,%s}: try_to_interpolate_data: did %d of %d elems\n\n",
+        Parallel::my_node(), Parallel::my_proc(), AhTag::label(), timestep,
+        interp_vars_info.at(timestep).element_has_interpolated.size(),
+        num_elements);
+  }
+  if (interp_vars_info.at(timestep).element_has_interpolated.size() ==
+      num_elements) {
+    send_data_to_ah_finder<Metavariables, DbTags, AhTag>(box, cache, timestep);
+  }
+}
+}  // namespace DataInterpolator_detail
+
+namespace Actions {
```

Comment:
again, split the actions into separate files, use the action testing framework,...